python/dgl/nn/pytorch/graph_transformer.py

```diff
@@ -336,3 +332,186 @@ def forward(self, g, edge_feat):
             path_encoding.append(sub_encoding)
 
         return th.stack(path_encoding, dim=0)
+
+
+class SpatialEncoder(nn.Module):
+    r"""Spatial Encoder, as introduced in
+    `Do Transformers Really Perform Bad for Graph Representation?
+    <https://proceedings.neurips.cc/paper/2021/file/f1c1592588411002af340cbaedd6fc33-Paper.pdf>`__
+    This module is a learnable spatial embedding module and encodes
+    the distance of the shortest path between each node pair as attention bias.
+
+    Parameters
+    ----------
+    max_dist : int
+        Upper bound of the shortest path distance
+        between each node pair to be encoded.
+        Each 2D distance will be clamped into the range :math:`[0, max_dist]`.
+    num_heads : int, optional
+        Number of attention heads if multi-head attention mechanism is applied.
+        Default : 1.
+
+    Examples
+    --------
+    >>> import torch as th
+    >>> import dgl
+    >>> from dgl.nn import SpatialEncoder
+
+    >>> u = th.tensor([0, 0, 0, 1, 1, 2, 3, 3])
+    >>> v = th.tensor([1, 2, 3, 0, 3, 0, 0, 1])
+    >>> g = dgl.graph((u, v))
+    >>> spatial_encoder = SpatialEncoder(2, num_heads=8)
+    >>> out = spatial_encoder(g)
+    """
+
+    def __init__(self, max_dist, num_heads=1):
+        super().__init__()
+        self.max_dist = max_dist
+        self.num_heads = num_heads
+        # deactivate node pair between which the distance is 0 or -1

```

Comment:
So node pairs with distance 0 or inf will be treated as the same?

Reply:
Sure, a pair of node and itself will be padded, the same with that in Graphormer and PathEncoder.